### PR TITLE
suggest a fix for precision on knob dragging

### PR DESCRIPTION
**Pull request by BaraMGB**
fixes https://github.com/LMMS/lmms/issues/3074 closed because it's a duplicate of:
fixes https://github.com/LMMS/lmms/issues/3024
~~only for knobs at first.~~

### CODE REVIEW

**Comment by BaraMGB** [comment]
Okay, this should be working. Can someone look at this?

**Comment by BaraMGB** [comment]
If no one is against this, I'll merge this within the next hours.

File: src/core/AutomatableModel.cpp [hunk 1]
```diff
@@ -275,9 +275,10 @@ float AutomatableModel::inverseScaledValue( float value ) const
 
 QString AutomatableModel::displayValue( const float val ) const
 {
+	const FloatModel *floatmodel = dynamic_cast<const FloatModel*>( this );
 	switch( m_dataType )
 	{
-		case Float: return QString::number( castValue<float>( scaledValue( val ) ) );
+		case Float: return QString::number( castValue<float>( scaledValue( floatmodel->getRoundedValue() ) ) );
```
Comment (grejppi) [inline src/core/AutomatableModel.cpp]: This function is used to convert any value into a string, not just the current value. `Reset` and `Paste value` are now incorrect:
image

Comment (BaraMGB) [inline src/core/AutomatableModel.cpp]: I have to look into it. Thank you for the hint!

Comment (BaraMGB) [inline src/core/AutomatableModel.cpp]: Yes, this is stupid. My intention was to fix the mixer bug ( 1.2345 % means 123,45 % ). I guess I revert this and search an other solution for this.

File: src/core/AutomatableModel.cpp [hunk 2]
```diff
@@ -713,6 +714,23 @@ float AutomatableModel::globalAutomationValueAt( const MidiTime& time )
 	}
 }
 
+float FloatModel::getRoundedValue() const
+{
+	return static_cast<float>( static_cast<int>( value() / step<float>() + 0.5 ) ) * step<float>();
```
Comment (Umcaruje) [inline src/core/AutomatableModel.cpp]: So this is causing a problem in the tripleoscillator coarse detune knobs: 
image
The negative values are off by one. 
And when you set it to zero
image
The actual value is -1 as you fixed this to use the true value in #3437 
image

Comment (BaraMGB) [inline src/core/AutomatableModel.cpp]: okay, got a fix for that. Is there an issue filed for? @Umcaruje

**Comment by Umcaruje** [comment]
Nope I just made this comment, never got around making an issue


On 1 Jun 2017 13:56, "Steffen Baranowsky" <notifications@github.com> wrote:

*@BaraMGB* commented on this pull request.
------------------------------

In src/core/AutomatableModel.cpp
<https://github.com/LMMS/lmms/pull/3075#discussion_r119594129>:

> @@ -713,6 +714,23 @@ float AutomatableModel::globalAutomationValueAt( const MidiTime& time )
 	}
 }

+float FloatModel::getRoundedValue() const
+{
+	return static_cast<float>( static_cast<int>( value() / step<float>()
+ 0.5 ) ) * step<float>();

okay, got a fix for that. Is there an issue filed for? @Umcaruje
<https://github.com/umcaruje>

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/LMMS/lmms/pull/3075#discussion_r119594129>, or mute the
thread

File: src/core/AutomatableModel.cpp [hunk 3]
```diff
@@ -713,6 +714,23 @@ float AutomatableModel::globalAutomationValueAt( const MidiTime& time )
...
+
 
 
+float FloatModel::getDigitCount()
+{
+	float steptemp = step<float>();
+	int digits = 0;
+	while ( steptemp < 1 )
+	{
+		steptemp = steptemp / 0.1f;
```
Comment (zapashcanon) [inline src/core/AutomatableModel.cpp]:
```c++
steptemp = steptemp * 10.0f; // easier to figure, and probably faster (but I guess the compiler already optimise it...)
```

Comment (BaraMGB) [inline src/core/AutomatableModel.cpp]: Thank you, I added this in this PR https://github.com/LMMS/lmms/pull/3597.

**Pull request merged**